Extension/src/LanguageServer/colorization.ts

```diff
@@ -0,0 +1,601 @@
+                            this.calculateThemeStyleForVsToken(TokenKind.UdlRaw, "constant.other.user-defined-literal", themeName, textMateRules);
+                            this.calculateThemeStyleForVsToken(TokenKind.GenericType, "entity.name.type.class.generic", themeName, textMateRules);
+                            return;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    public static createDecorationFromThemeStyle(themeStyle: ThemeStyle): vscode.TextEditorDecorationType {
+        if (themeStyle) {
+            if (themeStyle.foreground || themeStyle.background || themeStyle.fontStyle) {
+                let options: vscode.DecorationRenderOptions = {};
+                options.rangeBehavior = vscode.DecorationRangeBehavior.OpenOpen;
+                if (themeStyle.foreground) {
+                    options.color = themeStyle.foreground;
+                }
+                if (themeStyle.background) {
+                    options.backgroundColor = themeStyle.background;
+                }
+                if (themeStyle.fontStyle) {
+                    let parts: string[] = themeStyle.fontStyle.split(" ");
+                    parts.forEach((part) => {
+                        switch (part) {
+                            case "italic":
+                                options.fontStyle = "italic";
+                                break;
+                            case "bold":
+                                options.fontWeight = "bold";
+                                break;
+                            case "underline":
+                                options.textDecoration = "underline";
+                                break;
+                            default:
+                                break;
+                        }
+                    });
+                }
+                return vscode.window.createTextEditorDecorationType(options);
+            }
+        }
+
+        return null;
+    }
+}
+
+interface VersionedEdits {
+    editVersion: number;
+    changes: vscode.TextDocumentContentChangeEvent[];
+}
+
+export class ColorizationState {
+    private client: Client;
+    private colorizationSettings: ColorizationSettings;
+
+    private decorations: vscode.TextEditorDecorationType[] = new Array<vscode.TextEditorDecorationType>(TokenKind.Count);
+    private syntacticRanges: vscode.Range[][] = new Array<vscode.Range[]>(TokenKind.Count);
+    private semanticRanges: vscode.Range[][] = new Array<vscode.Range[]>(TokenKind.Count);
+
+    private inactiveDecoration: vscode.TextEditorDecorationType = null;
+    private inactiveRanges: vscode.Range[] = [];
+
+    private versionedEdits: VersionedEdits[] = [];
+    private lastSyntacticVersion: number = 0;
+    private lastSemanticVersion: number = 0;
+
+    public constructor(client: Client, colorizationSettings: ColorizationSettings) {
+        this.client = client;
```

Comment:
Is the RootUri only used for CppSettings? Do you need to create a new CppSettings everytime or can the one settings value that's queried be stored?

Reply:
As discussed, when creating CppSettings it calls vscode.workspace.getConfiguration(), which essentially takes a snapshot which does not change as settings change.  While OtherSettings calls vscode.workspace.getConfiguration each time a setting is accessed, it's inexpensive as it's just storing the Uri.  As a rule, I think we should create these settings objects as they are needed rather than keep them around.